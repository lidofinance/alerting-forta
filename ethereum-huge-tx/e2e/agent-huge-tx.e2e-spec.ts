import { configureContainer, Finding } from "forta-agent";
import { AwilixContainer, asFunction } from "awilix";
import {
  provideAgentPath,
  provideRunBlock,
  provideRunTransaction,
} from "./utils";

const TEST_TIMEOUT = 60_000; // ms

describe("agent-huge-tx e2e tests", () => {
  let runBlock: (
    blockHashOrNumber: string | number,
    initBlock?: number,
  ) => Promise<Finding[]>;
  let runTransaction: (txHash: string) => Promise<Finding[]>;
  let logSpy: jest.SpyInstance;
  let timeSpy: jest.SpyInstance;

  beforeAll(() => {
    logSpy = jest.spyOn(console, "log");
    logSpy.mockImplementation(() => {});
    timeSpy = jest.spyOn(Date, "now");
    timeSpy.mockImplementation(() => new Date("2023-12-31"));
  });

  beforeEach(async () => {
    const container = configureContainer() as AwilixContainer;
    container.register({
      agentPath: asFunction(
        provideAgentPath("subagents/huge-tx/agent-huge-tx"),
      ),
      runTransaction: asFunction(provideRunTransaction),
      runBlock: asFunction(provideRunBlock),
    });

    // https://docs.forta.network/en/latest/cli/#invoke-commands-programmatically
    runTransaction = container.resolve("runTransaction");
    runBlock = container.resolve("runBlock");
  });

  afterAll(() => {
    jest.resetAllMocks();
  });

  it(
    "should process block with huge change in Maker wstETH-B vault balance",
    async () => {
      const findings = await runBlock(15250794, 15250719);
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );

  it(
    "should process tx with huge token(s) transfer of Lido interest in a single TX",
    async () => {
      const findings = await runTransaction(
        "0x535d3848e0fd0715a455900d38816f57e7513663af24d7db0b1f6423cddc7821",
      );
      expect(findings.at(0)).toMatchSnapshot();
    },
    TEST_TIMEOUT,
  );
});
