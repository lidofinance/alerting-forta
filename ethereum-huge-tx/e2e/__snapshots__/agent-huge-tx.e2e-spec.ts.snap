// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`agent-huge-tx e2e tests should process block with huge change in Maker wstETH-B vault balance 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "HUGE-VAULT-BALANCE-CHANGE",
  "description": "**Maker wstETH-B** vault balance has **increased by 18.36%** during last 15 min.
Previous balance: 100848.31 wstETH
Current balance: 119359.57 wstETH
",
  "labels": Array [],
  "metadata": Object {},
  "name": "Huge change in Maker wstETH-B vault balance",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;

exports[`agent-huge-tx e2e tests should process tx with huge token(s) transfer of Lido interest in a single TX 1`] = `
Finding {
  "addresses": Array [],
  "alertId": "HUGE-TOKEN-TRANSFERS-IN-SINGLE-TX",
  "description": "**17.4k B-stETH-STABLE** were deposited to Aura Finance
by: 0x3f3429d28438cc14133966820b8a9ea61cf1d4f0 (unknown)
**17.4k B-stETH-STABLE** were deposited to Balancer Gauge
by: 0xaf52695e1bb01a16d33d7194c28c42b10e0dbec2 (Aura: Voter Proxy)
https://etherscan.io/tx/0x535d3848e0fd0715a455900d38816f57e7513663af24d7db0b1f6423cddc7821",
  "labels": Array [],
  "metadata": Object {},
  "name": "Huge token(s) transfer of Lido interest in a single TX",
  "protocol": "ethereum",
  "severity": 1,
  "type": 4,
}
`;
